### DISCUSSION TITLE
Default Interface Methods docs are a mess

### DISCUSSION THREAD

**Discussion by MgSam** [dotnet/csharplang#6956] [General]
There is a ton of confusion out in the community on how to use Default Interface Methods and the awful documentation makes matters worse. [Microsoft Learn/MSDN](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods#base-interface-invocations) has inherited the last version of the proposal from 2019 - which contains ideas like `override` that were not actually in the implementation as far as I can tell (the code under *Base interface invocations* does not compile). Mads' [blog post](https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/) from 4 years ago is similarly bereft of most details about the feature and should either be updated or deleted, IMHO.

In addition to just containing flat-out-wrong information, the docs also don't address the most common pit-of-failure with this feature which is that [people ](https://stackoverflow.com/questions/57761799/calling-c-sharp-interface-default-method-from-implementing-class) [expect ](https://github.com/dotnet/csharplang/discussions/6171) classes to inherit functionality from the interface.

``` C#
interface I { void Bark() => "woof"; }
class C : I { public void Greet() => Bark(); } //Error - Bark() is not visible unless `this` is cast to `I` first
```

Related, the docs should call out the fact that you cannot actually access a DIM from an implementing class unless it is static (this is something I think the language should find a solution for as the workaround is very clunky):
``` C#
interface I { void Bark() => "woof"; }
class C : I { public void Bark() => ((I)this).Bark(); } //Stack overflow here we come- no way to call a DIM once you've implemented the interface
```

Can only delegate to DIMs if you use statics.
``` C#
interface I { void Bark() => I.Bark(); static void Bark() => "woof"; }
class C : I { public void Bark() => I.Bark(); }
```

We absolutely need accurate docs for DIMs that include the current state of the feature, including the changes made in C# 11 around static members. You guys can't make fundamental type system changes and then not provide accurate documentation. Please fix this!

**Comment by CyrusNajmabadi** [comment]
@BillWagner Can you transfer this to the docs site.  I can't seem to do it on my end.  Thanks :)

**Reply by richlander** [reply]
Feel free to open up an issue at dotnet/docs that is more polite. We expect to discourse on these repos to be more reasonable.

This issue is not actionable as-is.
